### PR TITLE
[RISCV][MC] Implement evaluateBranch for auipc+jalr pairs

### PR DESCRIPTION
**Pull request by jobnoorman**
This patch implements `MCInstrAnalysis` state in order to be able analyze auipc+jalr pairs inside `evaluateBranch`.

This is implemented as follows:
- State: array of currently known GPR values;
- Whenever an auipc is detected in `updateState`, update the state value of RD with the immediate;
- Whenever a jalr is detected in `evaluateBranch`, check if the state holds a value for RS1 and use that to compute its target.

Note that this is similar to how binutils implements it and the output of llvm-objdump should now mostly match the one of GNU objdump.

This patch also updates the relevant llvm-objdump patches and adds a new one testing the output for interleaved auipc+jalr pairs.

This PR builds on #65479. Please only review the top commit here.

### CODE REVIEW

File: llvm/lib/Target/RISCV/MCTargetDesc/RISCVMCTargetDesc.cpp
Comment (arichardson) [inline llvm/lib/Target/RISCV/MCTargetDesc/RISCVMCTargetDesc.cpp:148]: This does not look safe to me, you will also have to invalidate the state for the destination register(s) for unhandled instructions otherwise you might report a wrong value.

File: llvm/test/tools/llvm-objdump/ELF/RISCV/multi-instr-target.s
Comment (arichardson) [inline llvm/test/tools/llvm-objdump/ELF/RISCV/multi-instr-target.s:10]: I'd also add a test here with an unhandled instruction e.g. mul between the auipc and a jalr that clobbers the destination register. Also need to be conservative for jumps etc since those could clobber registers (and we don't know the exact calling convention here since it could be hand-written assembly code)

**Review by arichardson — CHANGES_REQUESTED** [review]

File: llvm/lib/Target/RISCV/MCTargetDesc/RISCVMCTargetDesc.cpp
Comment (mtvec) [reply llvm/lib/Target/RISCV/MCTargetDesc/RISCVMCTargetDesc.cpp:173]: Indeed, it's probably better to be very conservative and make sure state is cleared for any instruction that may clobber registers. I've added a commit to implement this.

File: llvm/test/tools/llvm-objdump/ELF/RISCV/multi-instr-target.s
Comment (mtvec) [reply llvm/test/tools/llvm-objdump/ELF/RISCV/multi-instr-target.s]: Done.

File: llvm/lib/Target/RISCV/MCTargetDesc/RISCVMCTargetDesc.cpp
Comment (arichardson) [reply llvm/lib/Target/RISCV/MCTargetDesc/RISCVMCTargetDesc.cpp:173]: Thanks, the new change addresses my concerns.

**Comment by mtvec** [comment]
Ping.

**Comment by arichardson** [comment]
In general this looks good to me but I do worry a bit about the performance implications of regularly zeroing all the GPRS. Could you show the impact on disassembly of a large binary (e.g. statically linked clang for riscv?)

**Comment by mtvec** [comment]
> In general this looks good to me but I do worry a bit about the performance implications of regularly zeroing all the GPRS. Could you show the impact on disassembly of a large binary (e.g. statically linked clang for riscv?)

Here's a quick benchmark (release build without asserts):

```
$ file clang
clang: ELF 64-bit LSB pie executable, UCB RISC-V, RVC, double-float ABI, version 1 (GNU/Linux), dynamically linked, interpreter /lib/ld-linux-riscv64-lp64d.so.1, BuildID[sha1]=a4645a5d30617084df5efea9662d984d0a9dc918, for GNU/Linux 4.15.0, not stripped

$ size clang
     text	   data	    bss	      dec	    hex	filename
149308819	4260800	 622672	154192291	930c9a3	clang

$ hyperfine --parameter-list which main,pr './llvm-objdump.{which} -d clang > /dev/null' --warmup 3
Benchmark 1: ./llvm-objdump.main -d clang > /dev/null
  Time (mean ± σ):     56.003 s ±  0.206 s    [User: 32.285 s, System: 23.695 s]
  Range (min … max):   55.849 s … 56.511 s    10 runs

Benchmark 2: ./llvm-objdump.pr -d clang > /dev/null
  Time (mean ± σ):     56.651 s ±  0.071 s    [User: 32.911 s, System: 23.713 s]
  Range (min … max):   56.550 s … 56.797 s    10 runs

Summary
  ./llvm-objdump.main -d clang > /dev/null ran
    1.01 ± 0.00 times faster than ./llvm-objdump.pr -d clang > /dev/null
```

So there seems to be about 1% overhead.

If this is too much, one solution would be to not store an array of `std::optional<uint64_t>` but one containing just `uint64_t` and a separate 32-bit bitmap. I suppose that would remove most of the overhead of clearing state.

**Comment by arichardson** [comment]
> > In general this looks good to me but I do worry a bit about the performance implications of regularly zeroing all the GPRS. Could you show the impact on disassembly of a large binary (e.g. statically linked clang for riscv?)
> 
> Here's a quick benchmark (release build without asserts):
> 
> ```
> $ file clang
> clang: ELF 64-bit LSB pie executable, UCB RISC-V, RVC, double-float ABI, version 1 (GNU/Linux), dynamically linked, interpreter /lib/ld-linux-riscv64-lp64d.so.1, BuildID[sha1]=a4645a5d30617084df5efea9662d984d0a9dc918, for GNU/Linux 4.15.0, not stripped
> 
> $ size clang
>      text	   data	    bss	      dec	    hex	filename
> 149308819	4260800	 622672	154192291	930c9a3	clang
> 
> $ hyperfine --parameter-list which main,pr './llvm-objdump.{which} -d clang > /dev/null' --warmup 3
> Benchmark 1: ./llvm-objdump.main -d clang > /dev/null
>   Time (mean ± σ):     56.003 s ±  0.206 s    [User: 32.285 s, System: 23.695 s]
>   Range (min … max):   55.849 s … 56.511 s    10 runs
> 
> Benchmark 2: ./llvm-objdump.pr -d clang > /dev/null
>   Time (mean ± σ):     56.651 s ±  0.071 s    [User: 32.911 s, System: 23.713 s]
>   Range (min … max):   56.550 s … 56.797 s    10 runs
> 
> Summary
>   ./llvm-objdump.main -d clang > /dev/null ran
>     1.01 ± 0.00 times faster than ./llvm-objdump.pr -d clang > /dev/null
> ```
> 
> So there seems to be about 1% overhead.
> 
> If this is too much, one solution would be to not store an array of `std::optional<uint64_t>` but one containing just `uint64_t` and a separate 32-bit bitmap. I suppose that would remove most of the overhead of clearing state.

That seems better than expected and might be acceptable. However, I think using a bitset that can be zeroed with a single store should bring it down to near zero and will not add much complexity.

**Comment by mtvec** [comment]
> That seems better than expected and might be acceptable. However, I think using a bitset that can be zeroed with a single store should bring it down to near zero and will not add much complexity.

I implemented this and to my surprise, the overhead is still around 1%. After some digging, I noticed that the increase in runtime seems almost entirely due to the increased output size (total size increase is about 2% on the clang benchmark, the functions contributing the most to the increased runtime are output-related ones like `llvm::sys::unicode::columnWidthUTF8`).

Since this overhead is unavoidable, do you still feel it's worthwhile to implement the optimization?

**Comment by arichardson** [comment]
> > That seems better than expected and might be acceptable. However, I think using a bitset that can be zeroed with a single store should bring it down to near zero and will not add much complexity.
> 
> I implemented this and to my surprise, the overhead is still around 1%. After some digging, I noticed that the increase in runtime seems almost entirely due to the increased output size (total size increase is about 2% on the clang benchmark, the functions contributing the most to the increased runtime are output-related ones like `llvm::sys::unicode::columnWidthUTF8`).
> 
> Since this overhead is unavoidable, do you still feel it's worthwhile to implement the optimization?

I guess it makes sense that printing and looking up symbols is more expensive than zeroing 512 bytes somewhat regularly. I think the optimization does make sense unless it makes the code a lot less readable. But I'll leave that decision to other reviewers. The change looks good to me now but please wait for a review from someone else before merging.

**Review by arichardson — APPROVED** [review]

**Comment by mtvec** [comment]
> I think the optimization does make sense unless it makes the code a lot less readable.

I don't think it makes the code unreadable so I implemented the optimization.

**Pull request merged**
